### ISSUE TITLE
Questions about Reminders timing

### ISSUE THREAD

**Issue by yotamsh** [dotnet/orleans#4834]
Hey there,
In our team we are using Orleans Reminders to perform an action that should run exactly every 1 hour.
Reminders are registered with the following function: `RegisterOrUpdateReminder(string reminderName, TimeSpan dueTime, TimeSpan period);
`
As we noticed recently, a reminder that was registered with dueTime=0, period=1 hour, will not tick every one hour exactly.
It will simply wait 1 hour _after the last reminder ended_, and then send the next reminder tick.
That is a problem to us, since the reminder invokes a function that can take few seconds to end. In that case we will have an **accumulative delay** on reminder ticks, that will be received every 1 hour **+ epsilon**.

From [Orleans Documentation](https://dotnet.github.io/orleans/1.5/Documentation/Core-Features/Timers-and-Reminders.html), it is said to be the case for Orleans Timers, and for Reminders it said that "Reminders are similar to timers [...]".

My question is **whether there is an option to configure reminder timing** to be exactly as we registered (first tick after dueTime time, and then a tick every period), without being affected by the duration of the function the reminder invokes.
If there isn't, other solutions would be (A) making the reminder to invoke a async Task without awaiting for it, (B) making the reminder to start a timer to do the actual work, or (C) keeping track on the delay from the expected time and once it got too big update and re-schedule the reminder.

We'll be glad to hear if you know a way to handle it, or simply what would you do in such cases :)

**Comment by sergeybykov** [comment]
>It will simply wait 1 hour after the last reminder ended, and then send the next reminder tick.

Yes, this is by design - to prevent slow execution of reminder callback from clogging the 'queue' of next reminders. Timers work the same way.

>My question is whether there is an option to configure reminder timing to be exactly as we registered (first tick after dueTime time, and then a tick every period), without being affected by the duration of the function the reminder invokes.

There is no way to do this exactly.

>If there isn't, other solutions would be (A) making the reminder to invoke a async Task without awaiting for it, (B) making the reminder to start a timer to do the actual work, or (C) keeping track on the delay from the expected time and once it got too big update and re-schedule the reminder.

In general, the main purpose of reminders is fault tolerance - to make sure that the scheduled action will eventually execute, not precision of its timing. Latency of reminder execution is subject to many factors, the key one being current load on the system. Note that reminder ticks are generally delivered as remote messages *between* silos, which makes their timing dependent on current message queue sizes, etc.

Hence, option (A) only mitigates one piece of the latency equation. Option (B) is more robust because timers are local and much less expensive (as a result more precise) than reminders. So a timer scheduled relative to current local time will fire much closer to the desired time. I view (C) as just an optimization of the default case that only mitigates the problem.

If I were to do it, I would schedule a reminder to fire every 15 minutes, and have the callback check that a timer is scheduled for the next needed time. So three out of four reminder ticks would do nothing, only guarantee reliability in case of a failure recovery.

**Comment by philrowan-mtb** [comment]
I had a similar problem trying to build a cron like functionality using reminders. In the end our solution was to recalculate the `dueTime` after each execution based on the reminder received time. So if the reminder code takes time tR to run and we receive the reminder at t0 we update the reminder due time before exiting the `ReceiveReminder` method to `t0 + 1hr -  tR`. The issues about latency internal to the silo do still apply but the **accumulative delay** is mitigated. If it take 30 seconds to run for example it will be due again in 00:59:30 (1hr after start) rather than 00:60:00 (1hr and 30s after start).
